File: framework/languages/dotnet/samples/TicTacToe/Server/Play/Sessions/PlaySession.cs
using Systems.Zlink.Codecs.Json;
using TicTacToe.Server.Configuration;
using TicTacToe.Shared.Contracts;
using Zlink.Framework.Contracts.Actors;
using Zlink.Framework.Contracts.Channels;
using Zlink.Framework.Contracts.Streams;
using System.Text.Json;
using Systems.Zlink;
using Systems.Zlink.Stream.Connector.Contracts;

namespace TicTacToe.Server.Play.Sessions;

sealed class PlaySession(
    IZLinkSessionContext context,
    IZLinkActorManager actors,
    IZLinkChannelClient channels,
    ILogger<PlaySession> logger)
    : IZLinkSession
{
    private static readonly JsonSerializerOptions JsonOptions = new(JsonSerializerDefaults.Web);

    private string? _actorId;
    private IZLinkSessionActor? _actor;

    public IZLinkSessionContext Context { get; } = context;

    public ValueTask OnConnectedAsync(CancellationToken cancellationToken)
    {
        _ = cancellationToken;
        logger.LogInformation(
            "client -> play stream: connected. sessionId={SessionId}",
            Context.SessionId);
        return ValueTask.CompletedTask;
    }

    public ValueTask OnDisconnectedAsync(CancellationToken cancellationToken)
    {
        var actorId = _actorId;
        _actorId = null;
        _actor = null;
        logger.LogInformation(
            "client -> play stream: disconnected. sessionId={SessionId}, actor={ActorId}",
            Context.SessionId,
            actorId ?? "(unauthenticated)");

        _ = cancellationToken;
        return ValueTask.CompletedTask;
    }

    public ValueTask OnErrorAsync(
        ZLinkStreamError error,
        CancellationToken cancellationToken)
    {
        logger.LogWarning(
            "play stream: error. code={Code}, message={Message}, actor={ActorId}",
            error.Error,
            error.Diagnostic?.Message,
            _actorId ?? "(unauthenticated)");
        _ = cancellationToken;
        return ValueTask.CompletedTask;
    }

    public async ValueTask OnDispatchAsync(
        ZlinkStreamHeader header,
        Message payload,
        CancellationToken cancellationToken)
    {
        var actorId = _actorId;

        logger.LogInformation(
            "client -> play stream: message received. name={MessageName}, kind={Kind}, actor={ActorId}",
            header.Name,
            header.Kind,
            actorId ?? "(unauthenticated)");

        if (string.Equals(header.Name, nameof(AuthenticateReq), StringComparison.Ordinal))
        {
            var authenticated = await AuthenticateAsync(payload, cancellationToken);
            _actorId = authenticated.ActorId;
            _actor = authenticated.Actor;
            return;
        }

        if (actorId is null || _actor is not { } actor)
        {
            throw new InvalidOperationException("AuthenticateReq is required before play packets.");
        }

        logger.LogInformation(
            "play stream -> actor: dispatching packet. name={MessageName}, actor={ActorId}",
            header.Name,
            actorId);
        await actor.RelayAsync(header, payload, cancellationToken);
    }

    private async ValueTask<AuthenticatedPlaySession> AuthenticateAsync(
        Message payload,
        CancellationToken cancellationToken)
    {
        var authenticate = payload.Decode<AuthenticateReq>(JsonOptions);

        logger.LogInformation(
            "play stream -> api: authenticate requested. sessionId={SessionId}",
            Context.SessionId);

        var reply = await channels.RequestToChannel(
                SampleChannels.Api,
                new AuthenticatePlayerReq(authenticate.AccessToken))
            .Timeout(SampleTimeouts.Request)
            .SubmitAsync<AuthenticatePlayerRes>(cancellationToken);

        var playerActor = await actors.GetOrCreateAsync(
            reply.ActorId,
            SampleTypes.PlayerActor,
            cancellationToken);

        var actor = await Context.Actors.BindAsync(
            playerActor,
            cancellationToken);
        await Context.Client.Reply(new AuthenticateRes(reply.ActorId))
            .Submit();

        logger.LogInformation(
            "api -> play stream: authenticate accepted. sessionId={SessionId}, player={ActorId}, actor={ActorId}",
            Context.SessionId,
            reply.ActorId,
            actor.ActorId);

        return new AuthenticatedPlaySession(reply.ActorId, actor);
    }

    private readonly record struct AuthenticatedPlaySession(
        string ActorId,
        IZLinkSessionActor Actor);
}

File: framework/languages/node/samples/TicTacToe/Server/Play/Sessions/play-session.js
const { createChannelClient } = require('../../../../shared/channel-runtime');
const { PacketNames, SampleNames, SampleTimings } = require('../../../Shared/Contracts/messages');

class PlaySession {
  constructor(dependencies, transport) {
    this.dependencies = dependencies;
    this.transport = transport;
    this.actor = null;
  }

  async dispatch(header, payload) {
    if (header.name === PacketNames.authenticateReq) {
      await this.authenticate(header, payload);
      return;
    }
    if (this.actor === null) {
      throw new Error('AuthenticateReq is required before play packets.');
    }
    if (header.name === PacketNames.joinGameReq) {
      await this.joinGame(header, payload);
      return;
    }
    if (header.name === PacketNames.placeMarkReq) {
      await this.placeMark(header, payload);
      return;
    }
    throw new Error(`Unsupported play stream packet '${header.name}'.`);
  }

  async authenticate(header, request) {
    const apiClient = await createChannelClient({
      channelName: SampleNames.apiChannel,
      peers: [this.dependencies.apiEndpoint]
    });
    try {
      const authenticated = await apiClient
        .requestToChannel(SampleNames.apiChannel, { accessToken: request.accessToken })
        .packetName(PacketNames.authenticatePlayerReq)
        .timeout(SampleTimings.requestTimeout)
        .submit();
      this.actor = this.dependencies.actorFactory.ensure(authenticated.actorId);
      this.actor.displayName = authenticated.displayName;
      this.actor.session = this;
      this.transport.reply(header, {
        actorId: authenticated.actorId,
        displayName: authenticated.displayName
      });
    } finally {
      await apiClient.stop();
    }
  }

  async joinGame(header, request) {
    const result = this.dependencies.entrySpot.join(this.actor, request.gameId);
    this.transport.reply(header, result);
    await this.flushAllNotifications();
  }

  async placeMark(header, request) {
    const result = this.dependencies.placeMarkHandler.handle({
      actor: this.actor,
      gameId: request.gameId,
      cell: request.cell
    });
    this.transport.reply(header, result);
    await this.flushAllNotifications();
  }

  async flushAllNotifications() {
    for (const actor of this.dependencies.actorFactory.actors.values()) {
      if (actor.session !== null) {
        actor.session.flushNotifications(actor);
      }
    }
  }

  flushNotifications(actor) {
    while (actor.notifications.length > 0) {
      const notification = actor.notifications.shift();
      this.transport.send(notification.packetName, notification.payload, { seq: String(notification.seq) });
    }
  }
}

module.exports = { PlaySession };
